### DISCUSSION TITLE
Why don't we overload the (in)equality operators?

### DISCUSSION THREAD

**Discussion by ricardoV94** [pymc-devs/pytensor#84] [General]
```python
import numpy as np
np.arange(5) == 0  # array([ True, False, False, False, False])
```

```python
import pytensor.tensor as pt
pt.arange(5) == 0  # False
```

This has always been one of my main pet-peeves with Theano/Aesara. We can use all the same NumPy overloadad operators (`>`, `<`, `>=`, `~`, `&`, `^`) but not `==` and `!=`. I vaguely remember reading that the reason for this was that Theano-devs used equality for variable equality a lot in the internal code but that must certainly be better achieved with `is`, no?

```python
x = pt.arange(5)
x == x  # True
x == pt.arange(5)  # False
```

**Comment by twiecki** [comment]
Yeah, this seems wrong. Not much to add other than that your logic makes sense to me, and that I was surprised it differed from numpy in this this pretty central way.

**Comment by ricardoV94** [comment]
I found the answer to my question quickly enough: hashing: https://groups.google.com/g/theano-dev/c/AtQtU4E-L40/m/dJT-HH5b9WMJ

https://stackoverflow.com/a/46962289

**Reply by michaelosthege** [reply]
I don't quite follow these explanations.

When using vars as dictionary keys it is always about the object not the value. Whereever dict keys are compared by `==` it can and should be refactored to `is`, no?

**Reply by ricardoV94** [reply]
It doesn't really matter since we had to define equality by identity anyway. But we can change those if we find them.

What seems we can't really have in Python are hashable objects whose equality operator does "something fancy" for a lack of a better word.

The reason I opened the discussion was to understand why didn't we overload equality.
